### PR TITLE
fix: spec 0042, delayed LP fee

### PR DESCRIPTION
**Pull request by Jiajia-Cui**

### CODE REVIEW

**Review by davidsiska-vega — CHANGES_REQUESTED** [review]
I think we should go with the design that the delayed LP fees go into a per LP per market account and are paid out from there. 

It's no harder or easier than tracking a minimum (or average or another statistic) and it's cleaner. 

If rounding means that very small stakes don't get anything I view this as a feature not a bug (avoids having lots of tiny LPs).

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 1]
```diff
@@ -32,6 +32,12 @@ We now find the smallest integer `k` such that `[target stake] < sum from i=1 to
 
 Finally, we set the liquidity-fee-factor for this market to be the fee `LP-k-liquidity-fee-factor`. 
 
+## Delayed payout of LP fees
+
+For delayed payout of LP fees, the delayed LP fees go into a per LP per market account and are paid out from there.
+
+LPs with very small satkes will not get anything from rounding.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
```

Comment (Jiajia-Cui) [reply protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]: resolved by updating Distribution chapter

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 2]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
 **Warning** the above will be decimal calculations so the above checks will only be true up to a rounding errors.
 
 ### Distributing fees
-The liquidity fee is collected into either a per-market "bucket" belonging to liquidity providers for that market or into an account for each liquidity provider, according to their share of that fee. This account is not accessible by liquidity providers until the fee is distributed to them according to the below mechanism.
 
-We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `liquidity_providers_fee_distribition_time_step` which will define how frequently fees are distributed to a liquidity provider's general account for the market. 
+Once there is a trade happens, liquidity fee should be collected immediately into an account for each liquidity provider (we call it LP account) according to their share of that fee; therefore any LP provider who commited later than that (delayed LP) will not get the liquidity fee if there is no trade happened after the liquidity commitment has been made till the next `market.liquidity.providers.fee.distributionTimeStep`
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
Once there is a trade happens, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 3]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
 **Warning** the above will be decimal calculations so the above checks will only be true up to a rounding errors.
 
 ### Distributing fees
-The liquidity fee is collected into either a per-market "bucket" belonging to liquidity providers for that market or into an account for each liquidity provider, according to their share of that fee. This account is not accessible by liquidity providers until the fee is distributed to them according to the below mechanism.
 
-We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `liquidity_providers_fee_distribition_time_step` which will define how frequently fees are distributed to a liquidity provider's general account for the market. 
+Once there is a trade happens, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
On every trade, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 4]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
 **Warning** the above will be decimal calculations so the above checks will only be true up to a rounding errors.
 
 ### Distributing fees
-The liquidity fee is collected into either a per-market "bucket" belonging to liquidity providers for that market or into an account for each liquidity provider, according to their share of that fee. This account is not accessible by liquidity providers until the fee is distributed to them according to the below mechanism.
 
-We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `liquidity_providers_fee_distribition_time_step` which will define how frequently fees are distributed to a liquidity provider's general account for the market. 
+On every trade, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
+
+This account is not accessible by liquidity providers until the fee is distributed to them according to the below mechanism.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
This account is not accessible by liquidity providers until the fee is distributed to them according to the mechanism below.
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 5]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
...
 
-We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `liquidity_providers_fee_distribition_time_step` which will define how frequently fees are distributed to a liquidity provider's general account for the market. 
+On every trade, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
+
+This account is not accessible by liquidity providers until the fee is distributed to them according to the mechanism below.
+
+We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `market.liquidity.providers.fee.distributionTimeStep` which will define how frequently fees are distributed to a liquidity provider's general account for the market.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
A network parameter `market.liquidity.providers.fee.distributionTimeStep` will control how often fees are distributed from the LP fee account. Starting with the end of the opening auction the clock starts ticking and then rings every time `market.liquidity.providers.fee.distributionTimeStep` has passed. Every time this happens the fees are transferred to the liquidity provider's general account for the relevant asset. If `market.liquidity.providers.fee.distributionTimeStep` is set to `0` then the fees are distributed either immediately upon collection or at then end of a block. 
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 6]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
...
-The fee bucket contains `103.5 ETH`. We have `3` LPs with equity shares:
+We have `4` LPs with equity shares:
 share as below
 ```
 LP 1 eq share = 0.65
 LP 2 eq share = 0.25
 LP 3 eq share = 0.1
 ```
-When the time defined by ``liquidity_providers_fee_distribution_time_step` elapses we do transfers:
+Trade happened, and the fee bucket contains `103.5 ETH`. Liquidity fee should be collected immediately into the following LP account:
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
Trade happened, and the trade value for fee purposes multiplied by the liquidity fee factor is `103.5 ETH`. The following amounts be collected immediately into the LP fee accounts for the market:
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 7]
```diff
@@ -119,25 +119,42 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
...
+On every trade, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
+
+This account is not accessible by liquidity providers until the fee is distributed to them according to the mechanism below.
+
+A network parameter `market.liquidity.providers.fee.distributionTimeStep` will control how often fees are distributed from the LP fee account. Starting with the end of the opening auction the clock starts ticking and then rings every time `market.liquidity.providers.fee.distributionTimeStep` has passed. Every time this happens the fees are transferred to the liquidity provider's general account for the relevant asset. If `market.liquidity.providers.fee.distributionTimeStep` is set to `0` then the fees are distributed either immediately upon collection or at then end of a block.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
A network parameter `market.liquidity.providers.fee.distributionTimeStep` will control how often fees are distributed from the LP fee account. Starting with the end of the opening auction the clock starts ticking and then rings every time `market.liquidity.providers.fee.distributionTimeStep` has passed. Every time this happens the fees are transferred to the liquidity provider's margin account for the market. If `market.liquidity.providers.fee.distributionTimeStep` is set to `0` then the fees are distributed either immediately upon collection or at then end of a block. 
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 8]
```diff
@@ -167,4 +184,6 @@ When the time defined by ``liquidity_providers_fee_distribution_time_step` elaps
...
 - [ ] The total amount of liquidity fee distributed is equal to the most recent `liquidity-fee-factor` x `notional-value-of-all-trades` (<a name="0042-LIQF-011" href="#0042-LIQF-011">0042-LIQF-011</a>)
 - [ ] Liquidity providers with a commitment of 0 will not receive a share ot the fees (<a name="0042-LIQF-012" href="#0042-LIQF-012">0042-LIQF-012</a>)
-- [ ] If a market has `liquidity_providers_fee_distribution_time_step` set to more than `0` and such market settles then the fees are distributed as part of the settlement process, see [market lifecycle](./0043-MKTL-market_lifecycle.md). Any settled market has zero balance in the pool used to cumulate LP fees. (<a name="0042-LIQF-013" href="#0042-LIQF-013">0042-LIQF-013</a>)
+- [ ] If a market has `market.liquidity.providers.fee.distributionTimeStep` set to more than `0` and such market settles then the fees are distributed as part of the settlement process, see [market lifecycle](./0043-MKTL-market_lifecycle.md). Any settled market has zero balance in the pool used to cumulate LP fees. (<a name="0042-LIQF-013" href="#0042-LIQF-013">0042-LIQF-013</a>)
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
- [ ] If a market has `market.liquidity.providers.fee.distributionTimeStep` set to more than `0` and such market settles then the fees are distributed as part of the settlement process, see [market lifecycle](./0043-MKTL-market_lifecycle.md). Any settled market has zero balances in all the LP fee accounts. (<a name="0042-LIQF-014" href="#0042-LIQF-014">0042-LIQF-014</a>)
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 9]
```diff
@@ -120,11 +120,11 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
 
 ### Distributing fees
 
-Once there is a trade happens, liquidity fee should be collected immediately into an account for each liquidity provider (we call it LP account) according to their share of that fee; therefore any LP provider who commited later than that (delayed LP) will not get the liquidity fee if there is no trade happened after the liquidity commitment has been made till the next `market.liquidity.providers.fee.distributionTimeStep`
+On every trade, liquidity fee should be collected immediately into an account for each liquidity provider (call it LP fee account). Each party will have an LP fee account on every market on which they committed liquidity by providing LP stake. 
 
-This account is not accessible by liquidity providers until the fee is distributed to them according to the below mechanism.
+This account is not accessible by liquidity providers until the fee is distributed to them according to the mechanism below.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
This account is not under control of the LP party (they cannot initiate transfers in or out of the account). The account is under control of the network and funds from this account will be transferred to the owning LP party according to the mechanism below.
```

File: protocol/0042-LIQF-setting_fees_and_rewarding_lps.md [hunk 10]
```diff
@@ -120,11 +120,11 @@ There is a [Google sheet - requiring Vega login](https://docs.google.com/spreads
...
+This account is not under control of the LP party (they cannot initiate transfers in or out of the account). The account is under control of the network and funds from this account will be transferred to the owning LP party according to the mechanism below.
 
-We will create a new network parameter (which can be 0 in which case fees are transferred at the end of next block) called `market.liquidity.providers.fee.distributionTimeStep` which will define how frequently fees are distributed to a liquidity provider's general account for the market. 
+A network parameter `market.liquidity.providers.fee.distributionTimeStep` will control how often fees are distributed from the LP fee account. Starting with the end of the opening auction the clock starts ticking and then rings every time `market.liquidity.providers.fee.distributionTimeStep` has passed. Every time this happens the fees are transferred to the liquidity provider's margin account for the market. If `market.liquidity.providers.fee.distributionTimeStep` is set to `0` then the fees are distributed either immediately upon collection or at then end of a block.
```
Comment (davidsiska-vega) [inline protocol/0042-LIQF-setting_fees_and_rewarding_lps.md]:
```suggestion
A network parameter `market.liquidity.providers.fee.distributionTimeStep` will control how often fees are distributed from the LP fee account. Starting with the end of the opening auction the clock starts ticking and then rings every time `market.liquidity.providers.fee.distributionTimeStep` has passed. Every time this happens the balance in this account is transferred to the liquidity provider's margin account for the market. If `market.liquidity.providers.fee.distributionTimeStep` is set to `0` then the balance is distributed either immediately upon collection or at then end of a block. 
```

**Review by davidsiska-vega — APPROVED** [review]

**Pull request merged**